common/beerocks/bcl/source/beerocks_socket_event_loop.cpp

```diff
@@ -0,0 +1,359 @@
+                     << ") to the poll, but it's already there...";
+
+        return false;
+    }
+
+    // Build EventData structure
+    auto event_data      = std::make_shared<EventData>();
+    event_data->socket   = socket;
+    event_data->handlers = handlers;
+
+    // If timeout is set and a handler is defined, initialize a timerfd
+    if (timeout > std::chrono::milliseconds::zero()) {
+        if (handlers.on_timeout) {
+
+            // TODO: timerfd may not be supported on all kernels.
+            //       Implement a check at creation, and in case it's not supported
+            //       base the timeouts mechanism on the main timeout of the loop.
+
+            // Create a new timerfd file descriptor and initialize the timeout values
+            event_data->timeout_value = timeout;
+            event_data->timeout = std::make_shared<Socket>(timerfd_create(CLOCK_MONOTONIC, 0));
+
+            if (!event_data->timeout || event_data->timeout->getSocketFd() == -1) {
+                LOG(ERROR) << "Failed creating timerfd: " << strerror(errno);
+                return false;
+            }
+
+            // Convert the timeout into seconds and nano-seconds
+            auto timeout_sec = std::chrono::duration_cast<std::chrono::seconds>(timeout);
+            auto timeout_nanosec =
+                std::chrono::duration_cast<std::chrono::nanoseconds>(timeout - timeout_sec);
+
+            timespec timeout_spec({.tv_sec  = static_cast<time_t>(timeout_sec.count()),
+                                   .tv_nsec = static_cast<long>(timeout_nanosec.count())});
+            itimerspec timer_val({.it_interval = timeout_spec, .it_value = timeout_spec});
+
+            // Set the timerfd timeout value
+            if (timerfd_settime(event_data->timeout->getSocketFd(), 0, &timer_val, nullptr) == -1) {
+                LOG(ERROR) << "Failed setting timerfd value: " << strerror(errno);
+                return false;
+            }
+        } else {
+            LOG(WARNING) << "Timeout was requested for '" << timeout.count()
+                         << "' milliseconds, but not handler is provided. Ignoring.";
+        }
+    }
+
+    // Helper lambda function for adding a fd to the poll, and register for the following events:
+    // EPOLLIN: The associated fd is available for read operations.
+    // EPOLLOUT: The associated fd is available for write operations.
+    // EPOLLRDHUP: Socket peer closed connection, or shut down writing half of connection.
+    // EPOLLERR: Error condition happened on the associated fd.
+    // EPOLLHUP: Hang up happened on the associated fd.
+    auto add_fd_to_epoll = [&](int fd) -> bool {
+        epoll_event event = {};
+        event.data.fd     = fd;
+        event.events      = EPOLLIN | EPOLLRDHUP | EPOLLERR | EPOLLHUP;
+
+        // If an "on_write" handler is available, also register for write operations.
+        if (event_data->handlers.on_write) {
```

Comment:
If this one is conditional, then EPOLLIN should also be conditional, no?

Reply:
Yup, will do.

Reply:
Yup, will do.